metaflow/plugins/kfp/kfp_split_contexts.py

```diff
@@ -0,0 +1,189 @@
+            foreach_splits: <PASSED_IN_SPLIT_INDEXES>_index where index is ordinal of the split
+        }
+        """
+        if self.node.type == "foreach":
+            passed_in_split_indexes = os.environ[PASSED_IN_SPLIT_INDEXES_ENV_NAME]
+
+            # The splits are fed to kfp.ParallelFor to downstream steps as
+            # "passed_in_split_indexes" variable and become the step task_id
+            # Example: 0_3_1
+            foreach_splits = [
+                f"{passed_in_split_indexes}{SPLIT_INDEX_SEPARATOR}{split_index}".strip(
+                    SPLIT_INDEX_SEPARATOR
+                )
+                for split_index in range(0, flow._foreach_num_splits)
+            ]
+
+            return {
+                "task_id": current.task_id,
+                "foreach_splits": foreach_splits,
+            }
+        else:
+            return {
+                "task_id": current.task_id,
+            }
+
+    def get_parent_context(
+        self,
+        parent_context_step_name: str,
+        current_node: DAGNode,
+        passed_in_split_indexes: str,
+    ) -> Dict[str, str]:
+        """
+        Used by compute_input_paths() to access the parent context saved in step_kfp_output_path
+
+        Returns:
+            Task context dict built by build_context_dict() that was saved to S3 by kfp_decorator.
+        """
+        self.logger(parent_context_step_name, head="--")
+
+        context_node_task_id = str(
+            graph_to_task_ids(self.graph)[parent_context_step_name]
+        )
+        if self.graph[parent_context_step_name].is_inside_foreach:
+            if (
+                self.graph[parent_context_step_name].type == "foreach"
+                and parent_context_step_name
+                == current_node.in_funcs[0]  # and is a direct parent
+            ):
+                # if the direct parent node is a foreach
+                # and current_node.is_inside_foreach (we are in a foreach) then:
+                #   it's context_node_task_id is passed_in_split_indexes
+                #   minus the last split_index which is for the inner loop
+                split_indices_but_last_one = passed_in_split_indexes.split(
+                    SPLIT_INDEX_SEPARATOR
+                )[:-1]
+                context_split_indexes = SPLIT_INDEX_SEPARATOR.join(
+                    split_indices_but_last_one
+                )
+            else:
+                context_split_indexes = passed_in_split_indexes
+
+            context_node_task_id = f"{context_node_task_id}.{context_split_indexes}"
+        else:
+            # not is_inside_foreach, hence context_node_task_id is None
+            # and the step_kfp_output_path doesn't have task_id in it.
+            pass
+
+        self.logger(f"context_node_task_id: {context_node_task_id}")
+        step_kfp_output_path = self._build_step_kfp_output_path(
+            parent_context_step_name, context_node_task_id
+        )
+
+        with S3() as s3:
```

Comment:
Wondering if there is much of an overhead for this call to `S3`, I noticed in `kfp._compute_input_paths` we call `get_parent_context` many times so wondering if potentially it makes sense to create a `KfpSplitContext` in a context manager then it could handle opening/closing the S3 connection and reusing it between all the calls.

Might be a bit overkill if the overhead is low, I'm guessing it's low but just in case!